File: dotnet/Serpent.Test/CycleTest.cs
/// <summary>
/// Serpent, a Python literal expression serializer/deserializer
/// (a.k.a. Python's ast.literal_eval in .NET)
///
/// Copyright 2014, Irmen de Jong (irmen@razorvine.net)
/// Software license: "MIT software license". See http://opensource.org/licenses/MIT
/// </summary>

using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;

using NUnit.Framework;

namespace Razorvine.Serpent.Test
{
	[TestFixture]
	public class CycleTest
	{
		[Test]
		public void testTupleOk()
		{
			var ser = new Serializer();
			var t = new int[] {1,2,3};
			var d = new object[] {t,t,t};
			var data = ser.Serialize(d);
			var parser = new Parser();
	        var ast = parser.Parse(data);
		}

		[Test]
		public void testListOk()
		{
			var ser = new Serializer();
			var t = new List<int>();
			t.Add(1);
			t.Add(2);
			t.Add(3);
			var d = new List<Object>();
			d.Add(t);
			d.Add(t);
			d.Add(t);
			var data = ser.Serialize(d);
			var parser = new Parser();
	        var ast = parser.Parse(data);
		}

		[Test]
		public void testDictOk()
		{
			var ser = new Serializer();
			var t = new Hashtable();
			t["a"] = 1;
			var d = new Hashtable();
			d["x"] = t;
			d["y"] = t;
			d["z"] = t;
			var data = ser.Serialize(d);
			var parser = new Parser();
	        var ast = parser.Parse(data);
		}

		[Test]
		[Ignore("stackoverflow")]
		public void testListCycle()
		{
			var ser = new Serializer();
			var d = new List<Object>();
			d.Add(1);
			d.Add(2);
			d.Add(d);
			var data = ser.Serialize(d);
		}

		[Test]
		[Ignore("stackoverflow")]
		public void testDictCycle()
		{
			var ser = new Serializer();
			var d = new Hashtable();
			d["x"] = 1;
			d["y"] = 2;
			d["z"] = d;
			var data = ser.Serialize(d);
		}
		
		[Test]
		[Ignore("stackoverflow")]
		public void testClassCycle()
		{
			var ser = new Serializer();
			var d = new SerializeTestClass();
			d.x = 42;
			d.i = 99;
			d.s = "hello";
			d.obj = d;
			var data = ser.Serialize(d);
		}
	}
}

File: java/test/net/razorvine/serpent/test/CycleTest.java
/**
 * Serpent, a Python literal expression serializer/deserializer
 * (a.k.a. Python's ast.literal_eval in Java)
 * Software license: "MIT software license". See http://opensource.org/licenses/MIT
 * @author Irmen de Jong (irmen@razorvine.net)
 */
package net.razorvine.serpent.test;

import static org.junit.Assert.*;

import java.util.HashMap;
import java.util.LinkedList;

import net.razorvine.serpent.Parser;
import net.razorvine.serpent.Serializer;
import net.razorvine.serpent.ast.Ast;

import org.junit.Test;


public class CycleTest
{
	@Test
	public void testTupleOk()
	{
		Serializer ser = new Serializer();
		int[] t = new int[] {1,2,3};
		Object[] d = new Object[] {t,t,t};
		byte[] data = ser.serialize(d);
		Parser parser = new Parser();
        Ast ast = parser.parse(data);
	}

	@Test
	public void testListOk()
	{
		Serializer ser = new Serializer();
		LinkedList<Integer> t = new LinkedList<Integer>();
		t.add(1);
		t.add(2);
		t.add(3);
		LinkedList<Object> d = new LinkedList<Object>();
		d.add(t);
		d.add(t);
		d.add(t);
		byte[] data = ser.serialize(d);
		Parser parser = new Parser();
        Ast ast = parser.parse(data);
	}

	@Test
	public void testDictOk()
	{
		Serializer ser = new Serializer();
		HashMap<String, Integer> t = new HashMap<String, Integer>();
		t.put("a", 1);
		HashMap<String, Object> d = new HashMap<String, Object>();
		d.put("x", t);
		d.put("y", t);
		d.put("z", t);
		byte[] data = ser.serialize(d);
		Parser parser = new Parser();
        Ast ast = parser.parse(data);
	}

	@Test(expected=StackOverflowError.class)
	public void testListCycle()
	{
		Serializer ser = new Serializer();
		LinkedList<Object> d = new LinkedList<Object>();
		d.add(1);
		d.add(2);
		d.add(d);
		byte[] data = ser.serialize(d);
	}
	
	@Test(expected=StackOverflowError.class)
	public void testDictCycle()
	{
		Serializer ser = new Serializer();
		HashMap<String, Object> d = new HashMap<String, Object>();
		d.put("x", 1);
		d.put("y", 2);
		d.put("z", d);
		byte[] data = ser.serialize(d);
	}

	@Test(expected=StackOverflowError.class)
	public void testClassCycle()
	{
		Serializer ser = new Serializer();
		SerializationHelperClass thing = new SerializationHelperClass();
		thing.i = 42;
		thing.s = "hello";
		thing.x = 99;
		thing.obj = thing;
		byte[] data = ser.serialize(thing);
	}
}
